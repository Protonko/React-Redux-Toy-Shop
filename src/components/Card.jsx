import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Button} from './Common/Button';

export const Card = (product) => {
  const {price, sale, image, title, addToCart, countAddedProduct} = product;
  const productPrice = sale ? Math.round(price * sale) : price;
  const counterProduct = countAddedProduct ? `(${countAddedProduct})` : '';
  const classNamesPrice = classNames('card__data-price-text', {'card__data-price-text--new': sale});

  return (
    <article className="card">
      <div className="card__image">
        <img src={image} alt={title} className="card__image-pic"/>
      </div>

      <div className="card__data">
        <div className="card__section card__section--body">
          <div className="card__data-price">
            <span className={classNamesPrice}>
              ${productPrice}
            </span>
            {sale && (
              <s className="card__data-price-text card__data-price-text--old">
                ${price}
              </s>
            )}
          </div>
          <h2 className="card__data-title">
            {title}
          </h2>
        </div>
        <div className="card__section card__section--footer">
          <div className="card__data-actions">
            <ul className="card__data-actions-list list list--reset">
              <li className="card__data-actions-item">
                <Button
                  title={`Add to card ${counterProduct}`}
                  onClick={() => addToCart(product)}
                />
              </li>
            </ul>
          </div>
        </div>
      </div>
    </article>
  );
};

Card.propTypes = {
  price: PropTypes.number,
  sale: PropTypes.number,
  countAddedProduct: PropTypes.number,
  image: PropTypes.string,
  title: PropTypes.string,
  addToCart: PropTypes.func,
};
