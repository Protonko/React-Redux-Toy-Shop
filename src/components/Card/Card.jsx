import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Button} from '../Common/Buttons/Button';

export const Card = props => {
  const {product, selectedProducts, addToCart, removeFromCart} = props;
  const {price, sale, image, title} = product;
  const isAdded = selectedProducts.find(item => item.id === product.id);
  const productPrice = sale ? Math.round(price * sale) : price;
  const titleButton = isAdded ? 'Added' : 'Add to card';
  const classNamesPrice = classNames('card__data-price-text', {'card__data-price-text--new': sale});
  const classNamesButton = classNames('button', {'button--active': isAdded});

  const handleClickCartButton = product => {
    isAdded
      ? removeFromCart(product)
      : addToCart(product);
  };

  return (
    <article className="card">
      <div className="card__image">
        <img src={image} alt={title} className="card__image-pic"/>
      </div>

      <div className="card__data">
        <div className="card__section card__section--body">
          <div className="card__data-price">
            <span className={classNamesPrice}>
              ${productPrice}
            </span>
            {sale && (
              <s className="card__data-price-text card__data-price-text--old">
                ${price}
              </s>
            )}
          </div>
          <h2 className="card__data-title">
            {title}
          </h2>
        </div>
        <div className="card__section card__section--footer">
          <div className="card__data-actions">
            <ul className="card__data-actions-list list list--reset">
              <li className="card__data-actions-item">
                <Button
                  title={titleButton}
                  classNames={classNamesButton}
                  onClick={() => handleClickCartButton(product)}
                />
              </li>
            </ul>
          </div>
        </div>
      </div>
    </article>
  );
};

Card.propTypes = {
  product: PropTypes.object,
  selectedProducts: PropTypes.array,
  addToCart: PropTypes.func,
  removeFromCart: PropTypes.func,
};
