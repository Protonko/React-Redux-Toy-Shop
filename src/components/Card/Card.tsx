import React, {FC} from 'react';
import {Link} from 'react-router-dom';
import classNames from 'classnames';
import {IProduct} from 'typing/interfaces';
import {Button} from 'components/Common/Buttons/Button';

interface ICardProps {
  product: IProduct,
  selectedProducts: Array<IProduct>,
  addToCart: (product: IProduct) => any, // TODO - убрать
  removeFromCart: (product: IProduct) => any, // TODO - убрать
}

enum TitlesButton {
  ADDED = 'Added',
  ADD = 'Add to card',
}

type TTitleButton = TitlesButton.ADDED | TitlesButton.ADD

export const Card: FC<ICardProps> = props => {
  const {product, selectedProducts, addToCart, removeFromCart} = props;
  const {price, sale, image, title, id} = product;
  const productPrice: number = sale ? Math.round(price * sale) : price;
  const isAdded: boolean =
    !!selectedProducts.find((item: IProduct) => item.id === product.id);
  const titleButton: TTitleButton = isAdded ? TitlesButton.ADDED : TitlesButton.ADD;
  const classNamesPrice: string =
    classNames('card__data-price-text', {'card__data-price-text--new': sale});

  const handleClickCartButton = (product: IProduct): void => {
    isAdded ? removeFromCart(product) : addToCart(product);
  };

  return (
    <article className="card">
      <div className="card__image">
        <img src={image} alt={title} className="card__image-pic"/>
      </div>

      <div className="card__data">
        <div className="card__section card__section--body">
          <div className="card__data-price">
            <span className={classNamesPrice}>
              ${productPrice}
            </span>
            {sale && (
              <s className="card__data-price-text card__data-price-text--old">
                ${price}
              </s>
            )}
          </div>
          <Link to={`/product/${id}`} className="card__data-title">{title}</Link>
        </div>
        <div className="card__section card__section--footer">
          <div className="card__data-actions">
            <ul className="card__data-actions-list list list--reset">
              <li className="card__data-actions-item">
                <Button
                  title={titleButton}
                  classNamesAdditional={[{'button--active': isAdded}]}
                  onClick={() => handleClickCartButton(product)}
                />
              </li>
            </ul>
          </div>
        </div>
      </div>
    </article>
  );
};
