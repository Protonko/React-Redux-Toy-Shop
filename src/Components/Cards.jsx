import React from 'react';
import {Card} from './Card';

export const Cards = ({products}) => {
    return (
        <div className="cards">
            <ul className="cards__list list list--reset">
                {
                    products.map((product, index) => (
                        <li key={index} className="cards__item">
                            <Card {...product} />
                        </li>
                    ))
                }
            </ul>
        </div>
    );
};
