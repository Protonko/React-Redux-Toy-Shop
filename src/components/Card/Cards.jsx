import React from 'react';
import PropTypes from 'prop-types';
import Card from 'containers/Card';

export const Cards = ({products, selectedProducts}) => (
  <div className="cards">
    <ul className="cards__list list list--reset">
      {products.map((product, index) => (
        <li key={index} className="cards__item">
          <Card product={product} selectedProducts={selectedProducts} />
        </li>
      ))}
    </ul>
  </div>
);

Cards.propTypes = {
  products: PropTypes.object,
  selectedProducts: PropTypes.array
};
