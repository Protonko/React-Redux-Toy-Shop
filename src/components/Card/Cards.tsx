import React, {FC} from 'react';
import {IProduct} from 'models/interfaces';
import {Card} from 'components/Card/Card';

export interface IProps {
  products: Array<IProduct>,
  selectedProducts: Array<IProduct>,
  addToCart: (product: IProduct) => void,
  removeFromCart: (product: IProduct) => void,
}

export const Cards: FC<IProps> = ({
  products,
  selectedProducts,
  addToCart,
  removeFromCart
}) => {
  const renderCard = (product: IProduct) => (
    <li key={product.id} className="cards__item">
      <Card
        product={product}
        selectedProducts={selectedProducts}
        addToCart={addToCart}
        removeFromCart={removeFromCart}
      />
    </li>
  );

  return (
    <div className="cards">
      <ul className="cards__list list list--reset">
        {products.map(renderCard)}
      </ul>
    </div>
  );
}
