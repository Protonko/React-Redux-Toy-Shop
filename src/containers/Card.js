import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import * as cartActions from '../store/actions/cart';
import {Card} from '../components/Card/Card';

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators(cartActions, dispatch),
});

export default connect(null, mapDispatchToProps)(Card);
