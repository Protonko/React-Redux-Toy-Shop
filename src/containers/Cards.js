import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import * as cartActions from 'store/actions/cart';
import {Cards} from 'components/Card/Cards';

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators(cartActions, dispatch),
});

export default connect(null, mapDispatchToProps)(Cards);
