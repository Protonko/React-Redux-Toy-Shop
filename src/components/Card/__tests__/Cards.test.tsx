// types
import {IProps} from 'components/Card/Cards';

import React from 'react';
import {MemoryRouter} from 'react-router';
import {render} from 'enzyme';
import {products, selectedProducts} from 'static/__test__/card';
import {Cards} from 'components/Card/Cards';

const renderCardsComponent = (props: IProps) => (
  render(
    <MemoryRouter initialEntries={['/']}>
      <Cards {...props} />
    </MemoryRouter>
  )
)

describe('Cards', () => {
  let props: IProps;

  beforeEach(() => {
    props = {
      products,
      selectedProducts,
      addToCart: () => {},
      removeFromCart: () => {},
    };
  });

  it('Match snapshot', () => {
    const component = renderCardsComponent(props);

    expect(component).toMatchSnapshot();
  });
})
