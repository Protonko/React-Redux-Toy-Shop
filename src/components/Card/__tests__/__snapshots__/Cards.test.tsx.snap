// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards Match snapshot 1`] = `
<div
  class="cards"
>
  <ul
    class="cards__list list list--reset"
  >
    <li
      class="cards__item"
    >
      <article
        class="card"
      >
        <a
          class="card__image"
          href="/product/0"
        >
          <img
            alt="title"
            class="card__image-pic"
          />
        </a>
        <div
          class="card__data"
        >
          <div
            class="card__section card__section--body"
          >
            <div
              class="card__data-price"
            >
              <span
                class="card__data-price-text card__data-price-text--new"
              >
                $9
              </span>
              <s
                class="card__data-price-text card__data-price-text--old"
              >
                $10
              </s>
            </div>
            <a
              class="card__data-title"
              href="/product/0"
            >
              title
            </a>
          </div>
          <div
            class="card__section card__section--footer"
          >
            <div
              class="card__data-actions"
            >
              <ul
                class="card__data-actions-list list list--reset"
              >
                <li
                  class="card__data-actions-item"
                >
                  <button
                    class="button button--active"
                  >
                    Added
                  </button>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </article>
    </li>
    <li
      class="cards__item"
    >
      <article
        class="card"
      >
        <a
          class="card__image"
          href="/product/1"
        >
          <img
            alt="title 1"
            class="card__image-pic"
          />
        </a>
        <div
          class="card__data"
        >
          <div
            class="card__section card__section--body"
          >
            <div
              class="card__data-price"
            >
              <span
                class="card__data-price-text card__data-price-text--new"
              >
                $1
              </span>
              <s
                class="card__data-price-text card__data-price-text--old"
              >
                $1
              </s>
            </div>
            <a
              class="card__data-title"
              href="/product/1"
            >
              title 1
            </a>
          </div>
          <div
            class="card__section card__section--footer"
          >
            <div
              class="card__data-actions"
            >
              <ul
                class="card__data-actions-list list list--reset"
              >
                <li
                  class="card__data-actions-item"
                >
                  <button
                    class="button"
                  >
                    Add to card
                  </button>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </article>
    </li>
  </ul>
</div>
`;
